code/modules/antagonists/heretic/items/corrupted_organs.dm

```diff
@@ -0,0 +1,246 @@
+	var/amount_added = 5
+	/// What extra ingredients can we add?
+	var/list/extra_ingredients = list(
+		/datum/reagent/consumable/ethanol/pina_olivada,
+		/datum/reagent/consumable/ethanol/rum,
+		/datum/reagent/consumable/ethanol/thirteenloko,
+		/datum/reagent/consumable/ethanol/vodka,
+		/datum/reagent/consumable/superlaughter,
+		/datum/reagent/drug/bath_salts,
+		/datum/reagent/drug/blastoff,
+		/datum/reagent/drug/happiness,
+		/datum/reagent/drug/mushroomhallucinogen,
+	)
+
+/obj/item/organ/internal/liver/corrupt/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/corrupted_organ)
+
+/obj/item/organ/internal/liver/corrupt/on_mob_insert(mob/living/carbon/organ_owner, special)
+	. = ..()
+	RegisterSignal(organ_owner, COMSIG_ATOM_EXPOSE_REAGENTS, PROC_REF(on_drank))
+
+/obj/item/organ/internal/liver/corrupt/on_mob_remove(mob/living/carbon/organ_owner, special)
+	. = ..()
+	UnregisterSignal(organ_owner, COMSIG_ATOM_EXPOSE_REAGENTS)
+
+/// If we drank something, add a little extra
+/obj/item/organ/internal/liver/corrupt/proc/on_drank(atom/source, list/reagents, datum/reagents/source_reagents, methods)
+	SIGNAL_HANDLER
+	if (!(methods & INGEST))
+		return
+	var/datum/reagents/extra_reagents = new()
+	extra_reagents.add_reagent(pick(extra_ingredients), amount_added)
+	extra_reagents.trans_to(source, amount_added, transferred_by = src, methods = INJECT)
+
+
+/// Rapidly become hungry if you are not digesting blood
+/obj/item/organ/internal/stomach/corrupt
+	name = "corrupt stomach"
+	desc = "This parasite demands an unwholesome diet in order to be satisfied."
+	/// Do we have an unholy thirst?
+	var/thirst_satiated = FALSE
+	/// Timer for when we get thirsty again
+	var/thirst_timer
+
+/obj/item/organ/internal/stomach/corrupt/Initialize(mapload)
+	. = ..()
+	AddElement(/datum/element/corrupted_organ)
+	AddElement(/datum/element/noticable_organ, "%PRONOUN_They %PRONOUN_have an unhealthy pallor.")
+
+/obj/item/organ/internal/stomach/corrupt/on_mob_insert(mob/living/carbon/organ_owner, special)
+	. = ..()
+	RegisterSignal(organ_owner, COMSIG_ATOM_EXPOSE_REAGENTS, PROC_REF(on_drank))
+
+/obj/item/organ/internal/stomach/corrupt/on_mob_remove(mob/living/carbon/organ_owner, special)
+	. = ..()
+	UnregisterSignal(organ_owner, COMSIG_ATOM_EXPOSE_REAGENTS)
+
+/// Check if we drank a little blood
+/obj/item/organ/internal/stomach/corrupt/proc/on_drank(atom/source, list/reagents, datum/reagents/source_reagents, methods)
+	SIGNAL_HANDLER
+	if (!(methods & INGEST))
+		return
+
+	var/contains_blood = FALSE
+	for (var/datum/reagent in reagents)
+		if (!istype(reagent, /datum/reagent/blood))
+			continue
+		contains_blood = TRUE
+		break
+
+	if (!contains_blood)
+		return
+
+	if (!thirst_satiated)
+		to_chat(source, span_cult_italic("The thirst is satisfied... for now."))
+	thirst_satiated = TRUE
+	deltimer(thirst_timer)
+	thirst_timer = addtimer(CALLBACK(src, PROC_REF(start_thirsting)), 3 MINUTES, TIMER_STOPPABLE | TIMER_DELETE_ME)
```

Comment:
You can use `VARSET_CALLBACK` for this rather than needing a whole proc just to set the var back to `FALSE`